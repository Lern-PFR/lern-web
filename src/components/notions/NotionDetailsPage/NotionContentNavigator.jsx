import PropTypes from 'prop-types';
import _ from 'lodash';
import { ChevronLeft, ChevronRight } from 'react-feather';
import styled from 'styled-components';
import { StyledDiv, StyledList } from 'components/shared/layout';
import { navigationChevrons, navigator, stepperList } from 'theme/pages/notions/notionDetailsPage';

import NotionContentNavigatorStepper from './NotionContentNavigatorStepper';

// @TODO: Display the name of the document in a tooltip upon hovering a stepper.

const PreviousButton = styled(ChevronLeft)({ ...navigationChevrons });
const NextButton = styled(ChevronRight)({ ...navigationChevrons });

/**
 * @name NotionContentNavigator
 * @description A component used to navigate inside a lesson's pages.
 *
 * @author Timothée Simon-Franza
 *
 * @param {number}	currentDocOrder	The current document's order value.
 * @param {array}	notionContent	An array of the current notion's lessons and exercises.
 * @param {func}	redirectTo		Redirection method to trigger when a stepper is clicked.
 */
const NotionContentNavigator = ({ currentDocOrder, notionContent, redirectTo }) => {
	/**
	 * @name onPreviousClick
	 * @description onClick handler method for the 'previous' icon button.
	 *
	 * @author Timothée Simon-Franza
	 */
	const onPreviousClick = () => {
		if (currentDocOrder > 0) {
			redirectTo(currentDocOrder - 1);
		}
	};

	/**
	 * @name onNextClick
	 * @description onClick handler method for the 'next' icon button.
	 *
	 * @author Timothée Simon-Franza
	 */
	const onNextClick = () => {
		if (currentDocOrder < notionContent.length - 1) {
			redirectTo(currentDocOrder + 1);
		}
	};

	return (
		<StyledDiv {...navigator}>
			<PreviousButton role="button" data-testid="notion-navigatior-previous" onClick={onPreviousClick} disabled={currentDocOrder === 0} background="white" />
			<StyledList {...stepperList}>
				{notionContent.map(({ id, order, title = '', name = '' }) => (
					<NotionContentNavigatorStepper
						isCurrent={_.isEqual(notionContent[currentDocOrder]?.id, id)}
						key={id}
						label={title || name}
						onClick={() => redirectTo(order)}
					/>
				))}
			</StyledList>
			<NextButton role="button" data-testid="notion-navigatior-next" onClick={onNextClick} disabled={currentDocOrder === (notionContent.length - 1)} />
		</StyledDiv>
	);
};

NotionContentNavigator.propTypes = {
	currentDocOrder: PropTypes.number.isRequired,
	notionContent: PropTypes.arrayOf(
		PropTypes.oneOfType([
			PropTypes.shape({
				id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
				title: PropTypes.string,
				name: PropTypes.string,
				order: PropTypes.number.isRequired,
			}),
			PropTypes.shape({
				id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
				name: PropTypes.string.isRequired,
				order: PropTypes.number.isRequired,
			}),
		])
	).isRequired,
	redirectTo: PropTypes.func.isRequired,
};

export default NotionContentNavigator;
