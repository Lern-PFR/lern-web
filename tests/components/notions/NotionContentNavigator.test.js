import { shallow } from 'enzyme';
import { fireEvent, render, screen } from '@testing-library/react';
import NotionContentNavigator from 'components/notions/NotionDetailsPage/NotionContentNavigator';

describe('NotionContentNavigator', () => {
	const notionContent = [
		{ id: 0, name: 'test lesson 0', order: 0 },
		{ id: 1, name: 'test lesson 1', order: 1 },
		{ id: 2, name: 'test lesson 2', order: 2 },
		{ id: 3, title: 'test exercise 0', order: 3 },
		{ id: 4, name: 'test lesson 4', order: 4 },
	];

	it('should match previous snapshot', () => {
		const wrapper = shallow(<NotionContentNavigator currentDocOrder={0} notionContent={notionContent} redirectTo={jest.fn()} />);
		expect(wrapper).toMatchSnapshot();
	});

	it('should have the "previous" button disabled when the first document is displayed', () => {
		const redirectToSpy = jest.fn();
		render(<NotionContentNavigator currentDocOrder={0} notionContent={notionContent} redirectTo={redirectToSpy} />);
		fireEvent.click(screen.getByTestId('notion-navigatior-previous'));

		expect(redirectToSpy).toHaveBeenCalledTimes(0);
	});

	it('should have the "next" button disabled when the last document is displayed', () => {
		const redirectToSpy = jest.fn();
		render(<NotionContentNavigator currentDocOrder={4} notionContent={notionContent} redirectTo={redirectToSpy} />);
		fireEvent.click(screen.getByTestId('notion-navigatior-next'));

		expect(redirectToSpy).toHaveBeenCalledTimes(0);
	});

	it('should call the "redirectTo" method with the currentDocOrder - 1 when clicking on the "previous" button', () => {
		const redirectToSpy = jest.fn();
		render(<NotionContentNavigator currentDocOrder={3} notionContent={notionContent} redirectTo={redirectToSpy} />);
		fireEvent.click(screen.getByTestId('notion-navigatior-previous'));

		expect(redirectToSpy).toBeCalledTimes(1);
		expect(redirectToSpy).toHaveBeenCalledWith(2);
	});

	it('should call the "redirectTo" method with the currentDocOrder + 1 when clicking on the "next" button', () => {
		const redirectToSpy = jest.fn();
		render(<NotionContentNavigator currentDocOrder={3} notionContent={notionContent} redirectTo={redirectToSpy} />);
		fireEvent.click(screen.getByTestId('notion-navigatior-next'));

		expect(redirectToSpy).toBeCalledTimes(1);
		expect(redirectToSpy).toHaveBeenCalledWith(4);
	});
});
