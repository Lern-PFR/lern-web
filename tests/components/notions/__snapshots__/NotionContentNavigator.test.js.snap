// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotionContentNavigator should match previous snapshot 1`] = `
<styled.div
  alignItems="center"
  display="flex"
  flexDirection="row"
  justifyContent="center"
>
  <Styled(ChevronLeft)
    background="white"
    data-testid="notion-navigatior-previous"
    disabled={true}
    onClick={[Function]}
    role="button"
  />
  <styled.ul
    display="flex"
    flex={0.5}
    flexDirection="row"
    listStyle="none"
    marginLeft="1em"
    marginRight="1em"
    padding={0}
  >
    <NotionContentNavigatorStepper
      key="0"
      label="test lesson 0"
      onClick={[Function]}
    />
    <NotionContentNavigatorStepper
      key="1"
      label="test lesson 1"
      onClick={[Function]}
    />
    <NotionContentNavigatorStepper
      key="2"
      label="test lesson 2"
      onClick={[Function]}
    />
    <NotionContentNavigatorStepper
      key="3"
      label="test exercise 0"
      onClick={[Function]}
    />
    <NotionContentNavigatorStepper
      key="4"
      label="test lesson 4"
      onClick={[Function]}
    />
  </styled.ul>
  <Styled(ChevronRight)
    data-testid="notion-navigatior-next"
    disabled={false}
    onClick={[Function]}
    role="button"
  />
</styled.div>
`;
